pkg/controller/test-images-distributor/test_images_distributor.go

```diff
@@ -0,0 +1,364 @@
+	}
+
+	return imageStreamTag.Image.Name == reference.Image.Name, nil
+}
+
+const pullSecretName = "registry-cluster-pull-secret"
+
+func (r *reconciler) ensureImagePullSecret(namespace string, client ctrlruntimeclient.Client) error {
+	referenceSecret := r.generateReferencPullSecret(namespace)
+	name := types.NamespacedName{Namespace: namespace, Name: pullSecretName}
+	secret := &corev1.Secret{}
+	if err := client.Get(r.ctx, name, secret); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get secret %s: %w", name.String(), err)
+		}
+		// Tolerate IsAlreadyExist, another routine might have created it or our cache is stale
+		if err := client.Create(r.ctx, referenceSecret); err != nil && !apierrors.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create secret %s/%s: %v", referenceSecret.Namespace, referenceSecret.Name, err)
+		}
+	}
+
+	// Make it comparable by resetting the ObjectMeta fields the apiserver sets
+	resourceVersion := secret.ObjectMeta.ResourceVersion
+	secret.ObjectMeta = referenceSecret.ObjectMeta
+	if !apiequality.Semantic.DeepEqual(secret, referenceSecret) {
+		referenceSecret.ResourceVersion = resourceVersion
+		if err := client.Update(r.ctx, referenceSecret); err != nil {
+			return fmt.Errorf("failed to update secret %s/%s: %w", referenceSecret.Namespace, referenceSecret.Name, err)
+		}
+	}
+
+	return nil
+}
+
+func (r *reconciler) generateReferencPullSecret(namespace string) *corev1.Secret {
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      pullSecretName,
+		},
+		Data: map[string][]byte{
+			corev1.DockerConfigJsonKey: r.pullSecretGetter(),
+		},
+		Type: corev1.SecretTypeDockerConfigJson,
+	}
+}
+
+func testInputImageStreamTagFilterFactory(l *logrus.Entry, ca agents.ConfigAgent) (objectFilter, error) {
+	const indexName = "config-by-test-input-imagestreamtag"
+	if err := ca.AddIndex(indexName, indexConfigsByTestInputImageStramTag); err != nil {
+		return nil, fmt.Errorf("failed to add %s index to configAgent: %w", indexName, err)
+	}
+	l = logrus.WithField("subcomponent", "test-input-image-stream-tag-filter")
+	return func(nn types.NamespacedName) bool {
+		result, err := ca.GetFromIndex(indexName, nn.String())
+		// Today, GetFromIndex only errors if the index does not exist, so this should
+		// never happen.
+		if err != nil {
+			l.WithField("name", nn.String()).WithError(err).Error("Failed to get configs from index")
+			return false
+		}
+		return len(result) > 0
+	}, nil
+}
+
+func indexConfigsByTestInputImageStramTag(cfg api.ReleaseBuildConfiguration) []string {
+	m := apihelper.TestInputImageStreamTagsFromConfig(cfg)
+	var result []string
+	for key := range m {
+		result = append(result, key)
+	}
+	return result
+}
+
+func publicURLForImage(potentiallyPrivate string) string {
+	return strings.ReplaceAll(potentiallyPrivate, "docker-registry.default.svc:5000", "registry.svc.ci.openshift.org")
```

Comment:
this should realistically be configured with flags?

Reply:
Well, effectively this is a const for us

Reply:
I added a const for `svc.ci.openshift.org` -- maybe we can share it?